### PR TITLE
fix: update graviton allocatable memory calculation

### LINKED ISSUE #3937: Karpenter provisions under-sized instance for pending Pod
### Version

__Karpenter Version:__ 0.27.5

__Kubernetes Version:__ v1.27.1-eks-2f008fe

### Expected Behavior

When a Pod is unscheduled due to lack of capacity in k8s cluster, Karpenter brings up the right worker capacity and Pod is scheduled on the new node.

### Actual Behavior

Karpenter stucks in a loop that:

- It brings up a under-sized node, that has insufficient memory for the Pod + DaemonSets.
- Pod stays in unscheduled state.
- The node gets recycled by Karpenter due to being empty after TTL.
- Karpenter later on brings up another node, with the same under-sized spec, Pod stays unscheduled.

### Steps to Reproduce the Problem

See below manifests and logs section for details.

- Deploy a Pod to Kubernetes
- Inspect Kubernetes events and Karpenter logs, to verify that Karpenter is stuck in above "actual behaviour" loop of provisioning/deprovisioning under-sized instances.

### Resource Specs and Logs

(scheduling-unrelated fields are redacted)

#### Pod spec

The pod has 3 runtime-containers with total of `1408Mi` memory request

```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
  creationTimestamp: "2023-05-24T22:40:22Z"
  generateName: runner-dd9qghq1-project-431-concurrent-0
  labels:
    pod: runner-dd9qghq1-project-431-concurrent-0
    purpose: gitlab-runner-default
  name: runner-dd9qghq1-project-431-concurrent-0m8t6m
  namespace: gitlab-runner-default
  resourceVersion: "8666608"
  uid: 50decd30-93c7-4299-bad6-f94cb9431af8
spec:
  affinity: {}
  containers:
  - command: # redacted 
    image: # redacted 
    imagePullPolicy: Always
    name: build
    resources:
      requests:
        cpu: 500m
        memory: 1Gi
    securityContext: # redacted
    stdin: true
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts: # redacted 
  - command: # redacted 
    image: # redacted
    imagePullPolicy: Always
    name: helper
    resources:
      requests:
        cpu: 100m
        memory: 128Mi
    securityContext: # redacted 
    stdin: true
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts: # redacted
  - env: # redacted
    image: # redacted
    imagePullPolicy: Always
    name: svc-0
    resources:
      requests:
        cpu: 500m
        memory: 256Mi
    securityContext: # redacted
    stdin: true
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts: # redacted
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  hostAliases: # redacted
  imagePullSecrets:
  - name: runner-dd9qghq1-project-431-concurrent-0vxtfc
  initContainers:
  - command: # redacted
  nodeSelector:
    karpenter.sh/provisioner-name: gitlab-runner-default
    kubernetes.io/arch: arm64
    purpose: gitlab-runner-default
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  restartPolicy: Never
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 0
  tolerations:
  - effect: NoSchedule
    key: gitlab-runner-default
    operator: Exists
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes: # redacted
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2023-05-24T22:40:22Z"
    message: '0/8 nodes are available: 3 node(s) didn''t match Pod''s node affinity/selector,
      5 node(s) had untolerated taint {eks.amazonaws.com/compute-type: fargate}. preemption:
      0/8 nodes are available: 8 Preemption is not helpful for scheduling..'
    reason: Unschedulable
    status: "False"
    type: PodScheduled
  phase: Pending
  qosClass: Burstable
```

#### Karpenter logs

```
2023-05-24T22:40:23.094Z	DEBUG	controller.provisioner	101 out of 633 instance types were excluded because they would breach provisioner limits	{"commit": "698f22f-dirty"}
2023-05-24T22:40:23.107Z	INFO	controller.provisioner	found provisionable pod(s)	{"commit": "698f22f-dirty", "pods": 1}
2023-05-24T22:40:23.107Z	INFO	controller.provisioner	computed new machine(s) to fit pod(s)	{"commit": "698f22f-dirty", "machines": 1, "pods": 1}
2023-05-24T22:40:23.107Z	INFO	controller.provisioner	launching machine with 1 pods requesting {"cpu":"1225m","memory":"1408Mi","pods":"3"} from types r6g.large, r6gd.xlarge, m6gd.xlarge, m6gd.4xlarge, m7g.large and 44 other(s)	{"commit": "698f22f-dirty", "provisioner": "gitlab-runner-default"}
2023-05-24T22:40:23.725Z	DEBUG	controller.provisioner.cloudprovider	created launch template	{"commit": "698f22f-dirty", "provisioner": "gitlab-runner-default", "launch-template-name": "karpenter.k8s.aws/17890050121688119116", "launch-template-id": "lt-0fe0327b691ddc1ca"}
2023-05-24T22:40:25.988Z	INFO	controller.provisioner.cloudprovider	launched instance	{"commit": "698f22f-dirty", "provisioner": "gitlab-runner-default", "id": "i-09b20f6edc5454b85", "hostname": "ip-10-14-52-165.us-west-2.compute.internal", "instance-type": "t4g.small", "zone": "us-west-2c", "capacity-type": "spot", "capacity": {"cpu":"2","ephemeral-storage":"250Gi","memory":"1894Mi","pods":"11"}}
2023-05-24T22:41:13.089Z	DEBUG	controller.provisioner	101 out of 633 instance types were excluded because they would breach provisioner limits	{"commit": "698f22f-dirty"}
2023-05-24T22:41:13.101Z	INFO	controller.provisioner	found provisionable pod(s)	{"commit": "698f22f-dirty", "pods": 1}
2023-05-24T22:41:13.101Z	INFO	controller.provisioner	computed new machine(s) to fit pod(s)	{"commit": "698f22f-dirty", "machines": 1, "pods": 1}
2023-05-24T22:41:13.101Z	INFO	controller.provisioner	launching machine with 1 pods requesting {"cpu":"1225m","memory":"1408Mi","pods":"3"} from types r6g.large, r6gd.xlarge, m6gd.xlarge, m6gd.4xlarge, m7g.large and 44 other(s)	{"commit": "698f22f-dirty", "provisioner": "gitlab-runner-default"}
2023-05-24T22:41:15.493Z	INFO	controller.provisioner.cloudprovider	launched instance	{"commit": "698f22f-dirty", "provisioner": "gitlab-runner-default", "id": "i-0b575e8ff10706ca3", "hostname": "ip-10-14-52-135.us-west-2.compute.internal", "instance-type": "t4g.small", "zone": "us-west-2c", "capacity-type": "spot", "capacity": {"cpu":"2","ephemeral-storage":"250Gi","memory":"1894Mi","pods":"11"}}
2023-05-24T22:41:36.463Z	DEBUG	controller.node	added TTL to empty node	{"commit": "698f22f-dirty", "node": "ip-10-14-52-165.us-west-2.compute.internal"}
2023-05-24T22:41:53.099Z	DEBUG	controller.provisioner	101 out of 633 instance types were excluded because they would breach provisioner limits	{"commit": "698f22f-dirty"}
2023-05-24T22:41:53.116Z	INFO	controller.provisioner	found provisionable pod(s)	{"commit": "698f22f-dirty", "pods": 1}
2023-05-24T22:41:53.116Z	INFO	controller.provisioner	computed new machine(s) to fit pod(s)	{"commit": "698f22f-dirty", "machines": 1, "pods": 1}
2023-05-24T22:41:53.116Z	INFO	controller.provisioner	launching machine with 1 pods requesting {"cpu":"1225m","memory":"1408Mi","pods":"3"} from types r6g.large, r6gd.xlarge, m6gd.xlarge, m6gd.4xlarge, m7g.large and 44 other(s)	{"commit": "698f22f-dirty", "provisioner": "gitlab-runner-default"}
2023-05-24T22:41:57.485Z	INFO	controller.provisioner.cloudprovider	launched instance	{"commit": "698f22f-dirty", "provisioner": "gitlab-runner-default", "id": "i-019cf125b5616db7d", "hostname": "ip-10-14-52-163.us-west-2.compute.internal", "instance-type": "t4g.small", "zone": "us-west-2c", "capacity-type": "spot", "capacity": {"cpu":"2","ephemeral-storage":"250Gi","memory":"1894Mi","pods":"11"}}
2023-05-24T22:42:15.201Z	INFO	controller.deprovisioning	deprovisioning via emptiness delete, terminating 1 machines ip-10-14-52-165.us-west-2.compute.internal/t4g.small/spot	{"commit": "698f22f-dirty"}
2023-05-24T22:42:15.236Z	INFO	controller.termination	cordoned node	{"commit": "698f22f-dirty", "node": "ip-10-14-52-165.us-west-2.compute.internal"}
2023-05-24T22:42:15.242Z	DEBUG	controller.node	added TTL to empty node	{"commit": "698f22f-dirty", "node": "ip-10-14-52-135.us-west-2.compute.internal"}
2023-05-24T22:42:15.554Z	INFO	controller.termination	deleted node	{"commit": "698f22f-dirty", "node": "ip-10-14-52-165.us-west-2.compute.internal"}
2023-05-24T22:42:33.099Z	DEBUG	controller.provisioner	101 out of 633 instance types were excluded because they would breach provisioner limits	{"commit": "698f22f-dirty"}
2023-05-24T22:42:33.118Z	INFO	controller.provisioner	found provisionable pod(s)	{"commit": "698f22f-dirty", "pods": 1}
2023-05-24T22:42:33.118Z	INFO	controller.provisioner	computed new machine(s) to fit pod(s)	{"commit": "698f22f-dirty", "machines": 1, "pods": 1}
2023-05-24T22:42:33.118Z	INFO	controller.provisioner	launching machine with 1 pods requesting {"cpu":"1225m","memory":"1408Mi","pods":"3"} from types r6g.large, r6gd.xlarge, m6gd.xlarge, m6gd.4xlarge, m7g.large and 44 other(s)	{"commit": "698f22f-dirty", "provisioner": "gitlab-runner-default"}
2023-05-24T22:42:36.988Z	INFO	controller.provisioner.cloudprovider	launched instance	{"commit": "698f22f-dirty", "provisioner": "gitlab-runner-default", "id": "i-0b463e596717f3d2d", "hostname": "ip-10-14-52-154.us-west-2.compute.internal", "instance-type": "t4g.small", "zone": "us-west-2c", "capacity-type": "spot", "capacity": {"cpu":"2","ephemeral-storage":"250Gi","memory":"1894Mi","pods":"11"}}
2023-05-24T22:42:47.419Z	INFO	controller.deprovisioning	deprovisioning via emptiness delete, terminating 1 machines ip-10-14-52-135.us-west-2.compute.internal/t4g.small/spot	{"commit": "698f22f-dirty"}
```

#### Node Spec

From above Karpenter log, we can tell that Karpenter attempted to launch many `t4g.small` instances in response to the request of `"cpu":"1225m","memory":"1408Mi"`. But from the node spec, we see that the t4g.small nodes only have `1408312Ki` (~`1375Mi`) allocable memory.

```yaml
apiVersion: v1
kind: Node
metadata:
  annotations:
    alpha.kubernetes.io/provided-node-ip: 10.14.52.163
    node.alpha.kubernetes.io/ttl: "0"
    volumes.kubernetes.io/controller-managed-attach-detach: "true"
  creationTimestamp: "2023-05-24T22:41:57Z"
  finalizers:
  - karpenter.sh/termination
  labels:
    beta.kubernetes.io/arch: arm64
    beta.kubernetes.io/os: linux
    k8s.io/cloud-provider-aws: d2e133d6362a8b4e83cf62d5c97e1611
    karpenter.k8s.aws/instance-ami-id: ami-0ebbf0deba8bf7aba
    karpenter.k8s.aws/instance-category: t
    karpenter.k8s.aws/instance-cpu: "2"
    karpenter.k8s.aws/instance-encryption-in-transit-supported: "false"
    karpenter.k8s.aws/instance-family: t4g
    karpenter.k8s.aws/instance-generation: "4"
    karpenter.k8s.aws/instance-hypervisor: nitro
    karpenter.k8s.aws/instance-memory: "2048"
    karpenter.k8s.aws/instance-network-bandwidth: "128"
    karpenter.k8s.aws/instance-pods: "11"
    karpenter.k8s.aws/instance-size: small
    karpenter.sh/capacity-type: spot
    karpenter.sh/initialized: "true"
    karpenter.sh/provisioner-name: gitlab-runner-default
    kubernetes.io/arch: arm64
    kubernetes.io/hostname: ip-10-14-52-163.us-west-2.compute.internal
    kubernetes.io/os: linux
    node.kubernetes.io/instance-type: t4g.small
    purpose: gitlab-runner-default
    topology.kubernetes.io/region: us-west-2
    topology.kubernetes.io/zone: us-west-2c
  name: ip-10-14-52-163.us-west-2.compute.internal
  ownerReferences:
  - apiVersion: karpenter.sh/v1alpha5
    blockOwnerDeletion: true
    kind: Provisioner
    name: gitlab-runner-default
    uid: ec48bdce-6bea-4d76-b188-cfe685a10b85
  resourceVersion: "8667446"
  uid: 1d743662-7138-4c28-a418-85c3f059534e
spec:
  providerID: aws:///us-west-2c/i-019cf125b5616db7d
  taints:
  - effect: NoSchedule
    key: gitlab-runner-default
status:
  addresses:
  - address: 10.14.52.163
    type: InternalIP
  - address: ip-10-14-52-163.us-west-2.compute.internal
    type: Hostname
  allocatable:
    cpu: 1930m
    ephemeral-storage: "240498331658"
    hugepages-1Gi: "0"
    hugepages-2Mi: "0"
    hugepages-32Mi: "0"
    hugepages-64Ki: "0"
    memory: 1408312Ki
    pods: "11"
  capacity:
    cpu: "2"
    ephemeral-storage: 262122476Ki
    hugepages-1Gi: "0"
    hugepages-2Mi: "0"
    hugepages-32Mi: "0"
    hugepages-64Ki: "0"
    memory: 1895736Ki
    pods: "11"
  conditions:
  - lastHeartbeatTime: "2023-05-24T22:42:37Z"
    lastTransitionTime: "2023-05-24T22:42:37Z"
    message: kubelet is posting ready status
    reason: KubeletReady
    status: "True"
    type: Ready
  - lastHeartbeatTime: "2023-05-24T22:42:37Z"
    lastTransitionTime: "2023-05-24T22:42:29Z"
    message: kubelet has sufficient memory available
    reason: KubeletHasSufficientMemory
    status: "False"
    type: MemoryPressure
  - lastHeartbeatTime: "2023-05-24T22:42:37Z"
    lastTransitionTime: "2023-05-24T22:42:29Z"
    message: kubelet has no disk pressure
    reason: KubeletHasNoDiskPressure
    status: "False"
    type: DiskPressure
  - lastHeartbeatTime: "2023-05-24T22:42:37Z"
    lastTransitionTime: "2023-05-24T22:42:29Z"
    message: kubelet has sufficient PID available
    reason: KubeletHasSufficientPID
    status: "False"
    type: PIDPressure
  daemonEndpoints:
    kubeletEndpoint:
      Port: 10250
  images:
  - names:
    - 602401143452.dkr.ecr-fips.us-east-1.amazonaws.com/eks/pause:3.5
    - 602401143452.dkr.ecr-fips.us-east-2.amazonaws.com/eks/pause:3.5
    - 602401143452.dkr.ecr-fips.us-west-1.amazonaws.com/eks/pause:3.5
    - 602401143452.dkr.ecr-fips.us-west-2.amazonaws.com/eks/pause:3.5
    - 602401143452.dkr.ecr.af-south-1.amazonaws.com/eks/pause:3.5
    sizeBytes: 250452
  nodeInfo:
    architecture: arm64
    bootID: 0b887103-0690-4a22-bee6-b8c134b91037
    containerRuntimeVersion: containerd://1.6.19
    kernelVersion: 5.10.178-162.673.amzn2.aarch64
    kubeProxyVersion: v1.27.1-eks-2f008fe
    kubeletVersion: v1.27.1-eks-2f008fe
    machineID: ec2d7a7ea7f7d79f0fda92c2095f227b
    operatingSystem: linux
    osImage: Amazon Linux 2
    systemUUID: ec2d7a7e-a7f7-d79f-0fda-92c2095f227b
```

#### Karpenter Objects

```yaml
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"karpenter.sh/v1alpha5","kind":"Provisioner","metadata":{"annotations":{},"name":"gitlab-runner-default"},"spec":{"kubeletConfiguration":{"clusterDNS":["10.100.0.10"]},"labels":{"purpose":"gitlab-runner-default"},"limits":{"resources":{"cpu":100}},"providerRef":{"name":"gitlab-runner-default"},"requirements":[{"key":"karpenter.k8s.aws/instance-category","operator":"In","values":["a","m","t","r","c"]},{"key":"karpenter.k8s.aws/instance-cpu","operator":"Lt","values":["17"]},{"key":"karpenter.k8s.aws/instance-hypervisor","operator":"In","values":["nitro"]},{"key":"kubernetes.io/os","operator":"In","values":["linux"]},{"key":"karpenter.sh/capacity-type","operator":"In","values":["spot"]},{"key":"kubernetes.io/arch","operator":"In","values":["amd64","arm64"]}],"taints":[{"effect":"NoSchedule","key":"gitlab-runner-default"}],"ttlSecondsAfterEmpty":30,"ttlSecondsUntilExpired":86400}}
  creationTimestamp: "2023-05-02T16:52:06Z"
  generation: 5
  name: gitlab-runner-default
  resourceVersion: "8673924"
  uid: ec48bdce-6bea-4d76-b188-cfe685a10b85
spec:
  kubeletConfiguration:
    clusterDNS:
    - 10.100.0.10
  labels:
    purpose: gitlab-runner-default
  limits:
    resources:
      cpu: "100"
  providerRef:
    name: gitlab-runner-default
  requirements:
  - key: karpenter.k8s.aws/instance-category
    operator: In
    values:
    - a
    - m
    - t
    - r
    - c
  - key: karpenter.k8s.aws/instance-cpu
    operator: Lt
    values:
    - "17"
  - key: karpenter.k8s.aws/instance-hypervisor
    operator: In
    values:
    - nitro
  - key: kubernetes.io/os
    operator: In
    values:
    - linux
  - key: karpenter.sh/capacity-type
    operator: In
    values:
    - spot
  - key: kubernetes.io/arch
    operator: In
    values:
    - amd64
    - arm64
  taints:
  - effect: NoSchedule
    key: gitlab-runner-default
  ttlSecondsAfterEmpty: 30
  ttlSecondsUntilExpired: 86400
status:
  resources:
    cpu: "2"
    ephemeral-storage: 262131692Ki
    memory: 950104Ki
    pods: "4"
```

```yaml
apiVersion: karpenter.k8s.aws/v1alpha1
kind: AWSNodeTemplate
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration:  # redacted
  creationTimestamp: "2023-05-02T16:52:05Z"
  generation: 4
  name: gitlab-runner-default
  resourceVersion: "8671558"
  uid: fbc4ae1a-15e5-4410-aa35-2fbca0fad617
spec:
  blockDeviceMappings:
  - deviceName: /dev/xvda
    ebs:
      encrypted: true
      iops: 3000
      throughput: 125
      volumeSize: 250Gi
      volumeType: gp3
  securityGroupSelector:
    karpenter.sh/discovery/gitlab-runner-on-eks-green: gitlab-runner-on-eks-green
  subnetSelector: # redacted
  tags: # redacted
  userData: |
    MIME-Version: 1.0
    Content-Type: multipart/mixed; boundary="BOUNDARY"

    --BOUNDARY
    Content-Type: text/x-shellscript; charset="us-ascii"

    #!/bin/bash
    echo "$(jq '.registryPullQPS=30' /etc/kubernetes/kubelet/kubelet-config.json)" > /etc/kubernetes/kubelet/kubelet-config.json
    echo "$(jq '.registryBurst=50' /etc/kubernetes/kubelet/kubelet-config.json)" > /etc/kubernetes/kubelet/kubelet-config.json

    --BOUNDARY--
status:
  securityGroups: # redacted
  subnets:  # redacted
```

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by tzneal**
Fixes #3937 3937

**Description**

We are calculating slightly too much allocatable memory for some graviton instances.  Subtract an extra 64MiB from graviton only to account for cma reserved memory.

**How was this change tested?**

* Unit testing & on EKS

**Does this change impact docs?**
- [X] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [ ] No

**Release Note**

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/providers/instancetype/types.go [hunk 1]
```diff
@@ -162,7 +162,12 @@ func cpu(info *ec2.InstanceTypeInfo) *resource.Quantity {
 }
 
 func memory(ctx context.Context, info *ec2.InstanceTypeInfo) *resource.Quantity {
-	mem := resources.Quantity(fmt.Sprintf("%dMi", *info.MemoryInfo.SizeInMiB))
+	sizeInMib := *info.MemoryInfo.SizeInMiB
+	// Gravitons have an extra 65 MiB of cma reserved memory that we can't use
```
Comment (jonathan-innis) [inline pkg/providers/instancetype/types.go:166]:
```suggestion
	// Gravitons have an extra 64 MiB of cma reserved memory that we can't use
```

File: pkg/providers/instancetype/suite_test.go [hunk 2]
```diff
@@ -1048,6 +1056,57 @@ var _ = Describe("Instance Types", func() {
...
+			Expect(err).To(BeNil())
+
+			instanceTypes := map[string]*corecloudprovider.InstanceType{}
+			for _, it := range its {
+				instanceTypes[it.Name] = it
+			}
+
+			for _, tc := range []struct {
+				InstanceType string
+				// Actual allocatable values as reported by the node from kubelet. You find these
```
Comment (jonathan-innis) [inline pkg/providers/instancetype/suite_test.go:1072]: Nice 🎉

**Review by jonathan-innis — APPROVED** [review]
One nit, but otherwise LGTM

File: pkg/providers/instancetype/types.go (hunk 1 above)
Comment (tzneal) [reply pkg/providers/instancetype/types.go]: Ah, nice, fixed!

**Review by jonathan-innis — APPROVED** [review]
LGTM 🚀

**Pull request merged**